Objects/abstract.c

```diff
@@ -2147,55 +2147,77 @@ PyMapping_HasKey(PyObject *o, PyObject *key)
     return 0;
 }
 
+/* This function is quite similar to PySequence_Fast(), but specialized to be
+   a helper for PyMapping_Keys(), PyMapping_Items() and PyMapping_Values().
+ */
+static PyObject *
+method_output_as_list(PyObject *o, _Py_Identifier *meth_id)
+{
+    PyObject *it, *result, *meth_output;
+
+    assert(o != NULL);
+    meth_output = _PyObject_CallMethodId(o, meth_id, NULL);
+    if (meth_output == NULL || PyList_CheckExact(meth_output)) {
+        return meth_output;
+    }
+    it = PyObject_GetIter(meth_output);
```

Comment:
Why not just calling list(obj)?

Reply:
`PySequence_List()` few lines below is just a C API equivalent of calling `list()`. But first we want to produce better error message if `keys()` returns non-iterable. It is in the current code, and removing it will simplify the code, but introduce a regression.

Reply:
I would prefer to pass meth_output to PySequence_List() rather than iter(meth_output). list_extend() has a fast-path for tuple.

Reply:
This doesn't matter. `keys()` returning tuple should be very rare. I don't know any example.

Calling `PyObject_GetIter()` on an iterator is cheap, it should return the iterator itself. But calling `PyObject_GetIter()` on a general iterable creates a new iterator object. This is not so cheap. This will add an overhead (and possible side effects) in most common cases.

Reply:
list(iter(obj)) is not exactly the same than list(obj). My point on performance was the cost of PySequence_List(iterator) vs PySequence_List(tuple).

Reply:
We could add a special case for tuples:
```
if (PyTuple_CheckExact(meth_output)) {
    result = PySequence_List(meth_output);
    Py_DECREF(meth_output);
    return result;
}
```
But I think this is a premature optimization. Do you know any class that return `keys()` as a tuple?

Reply:
Nah, it's fine. I don't care much. The current code is ok.